### DISCUSSION TITLE
TS: Make incorrect `RouteName`s a Type Error?

### DISCUSSION THREAD

**Discussion by studnitz** [tighten/ziggy#708] [General]
Hi there,

I'm really happy about the added TypeScript support. 
I wanted to ask if it would be possible to make the `route()` function to only accept valid `RouteName` and then error if it's not in the route list?
This is the case for example for [`momentum-trail`](https://github.com/lepikhinb/momentum-trail-helper), which is a Ziggy wrapper.

Having TypeScript yell at you, when you had a typo in the route name or when you refactored some names was really helpful in that case. :)

**Comment by studnitz** [comment]
We managed to self roll this, if anyone is interested:

```ts
import ziggyRoute from 'ziggy-js'

import type { RouteName, RouteParams } from 'ziggy-js'

type RemoveStringFromUnion<T> = T extends T ? (string & {} extends T ? never : T) : never

export type StrictRouteName = RemoveStringFromUnion<RouteName>

export function route<T extends StrictRouteName>(name: T, parameters?: RouteParams<T>, absolute?: boolean) {
  return ziggyRoute(name, parameters, absolute)
}
```

**Comment by bakerkretzmar** [comment]
I really like this idea but I have this nagging feeling that there's _some_ reason I can't think of that we need to allow strings that we might think aren't valid route names 🙃

I wonder if there's any way to make this strictness optional? In the meantime your solution is really clean, and if we implement this I'm going to come back and refer to it.

Thanks!

**Reply by studnitz** [reply]
I understand your concern. In that case the user would have still the ability to add a `// @ts-expect-error` to silence that error.

And for the optional strictness: As you are generating the types anyways you could also generate the `RouteName` definition based on some `strict_types` config value in `ziggy.php` or CLI param?
